### PR TITLE
[#5104] Implement sending report via email feature

### LINKED ISSUE #5104: Feature Request: Send program report via email
### What are you trying to do?

Generating program reports can take a long time especially if it is a large program with a lot of data.

### Describe the solution you'd like

Instead of downloading it right away, generate the reports in the background process and email them when it's done.

### Have you consider alternatives?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by zuhdil**
# TODO / Done

Summarize what has been changed / what has to be done in order to finalize the PR.

 - [x] Change program report endpoint to queue report generation instead of download
 - [x] Generate program report in the background process and send it via email 
 - [x] Handle report queueing in the frontend

# Test plan

What tests are necessary to ensure this works or doesn't break anything working

 - [x] Unit test
 - [x] Manual test
 
## Manual testing on local dev

 - Open the program report tab
 - Click on a report
 - A popup notification will be shown
 - Go to terminal and run `docker-compose exec -u akvo web ./manage.py send_report_via_email`
 - The report will be generated and sent via email

---

Closes #5104

### CODE REVIEW

File: akvo/rsr/views/py_reports/nuffic_country_level_map_report.py [hunk 1]
```diff
@@ -19,6 +19,68 @@
...
+    show_comment = True if params.get('comment', '') == 'true' else False
+    start_date = utils.parse_date(params.get('period_start', ''), datetime(1900, 1, 1))
+    end_date = utils.parse_date(params.get('period_end', ''), datetime(2999, 12, 31))
+
+    country = Country.objects.get(iso_code=country)
+    program = Project.objects.get(pk=params['program_id'])
+
+    project_ids = program.descendants()\
+        .exclude(pk=program.id)\
+        .exclude(Q(title__icontains='test') | Q(subtitle__icontains='test'))\
```
Comment (MichaelAkvo) [inline akvo/rsr/views/py_reports/nuffic_country_level_map_report.py:58]: :thinking: does nuffic only have one program and it will never contain test?

**Review by MichaelAkvo — CHANGES_REQUESTED** [review]

File: akvo/rsr/views/py_reports/email_report.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+def _get_pending_jobs():
+    started_timeout = now() - TIMEOUT
+    return EmailReportJob.objects\
+        .order_by('created_at')\
+        .filter(finished_at__isnull=True)\
+        .exclude(Q(attempts__gte=MAX_ATTEMPTS) | Q(started_at__gte=started_timeout))
+
+
+def _get_report_handler(report):
+    return HANDLER[report] if report in HANDLER else None
```
[the remark this answers is not in the document]
Comment (zuhdil) [reply akvo/rsr/views/py_reports/email_report.py]: Fixed

File: akvo/rsr/views/py_reports/results_indicators_with_map_pdf_reports.py [hunk 3]
```diff
@@ -19,6 +19,71 @@
...
+            'start_date': start_date,
+            'end_date': end_date,
+        },
+        recipient=request.user.email
+    )
+
+
+def handle_org_projects_email_report(params, recipient):
+    country = params.get('country')
+    show_comment = True if params.get('comment', '') == 'true' else False
```
[the remark this answers is not in the document]
Comment (zuhdil) [reply akvo/rsr/views/py_reports/results_indicators_with_map_pdf_reports.py]: Fixed

File: akvo/rsr/views/py_reports/results_indicators_with_map_pdf_reports.py [hunk 4]
```diff
@@ -19,6 +19,71 @@
...
+        'projects',
+        'projects__results',
+        'projects__results__indicators',
+        'projects__results__indicators__periods'
+    ).get(pk=project_hierarchy.organisation.id)
+    projects = organisation.all_projects().filter(primary_location__country=country)
+    coordinates = [
+        Coordinate(p.primary_location.latitude, p.primary_location.longitude)
+        for p
+        in projects
```
[the remark this answers is not in the document]
Comment (zuhdil) [reply akvo/rsr/views/py_reports/results_indicators_with_map_pdf_reports.py]: Fixed

File: akvo/rsr/views/py_reports/utils.py [hunk 5]
```diff
@@ -16,10 +16,50 @@
...
+from akvo.utils import ObjectReaderProxy, ensure_decimal, rsr_send_mail
+
+
+def add_email_report_job(report, payload, recipient):
+    EmailReportJob.objects.create(
+        report=report,
+        payload=payload,
+        recipient=recipient
+    )
+    return HttpResponse('Your report is being prepared. It will be sent to your email in a few moments.', status=202)
```
[the remark this answers is not in the document]
Comment (zuhdil) [reply akvo/rsr/views/py_reports/utils.py]: Fixed

File: akvo/rsr/views/py_reports/nuffic_country_level_map_report.py (hunk 1 above)
Comment (zuhdil) [reply akvo/rsr/views/py_reports/nuffic_country_level_map_report.py:58]: At the time this feature was requested, nuffic had several test projects that they didn't want to show in the report. So far, there has been no request from them to negate it.

The query should work even if nuffic has more than one program. It should also work for other partner programs if enabled.

**Pull request merged**